AsyncDisplayKit/Details/ASRangeController.mm

```diff
@@ -105,53 +105,44 @@ - (void)updateCurrentRangeWithMode:(ASLayoutRangeMode)rangeMode
     _currentRangeMode = rangeMode;
     _didUpdateCurrentRange = YES;
     
-    [self scheduleRangeUpdate];
+    [self performRangeUpdateSynchronously:YES];
   }
 }
 
-- (void)scheduleRangeUpdate
+- (void)performRangeUpdateSynchronously:(BOOL)updateSynchronously
 {
-  if (_queuedRangeUpdate) {
-    return;
+  if (updateSynchronously) {
+    ASDisplayNodeAssertMainThread();
+    [self _updateVisibleNodeIndexPaths];
+  } else if (!_queuedRangeUpdate) {
+    _queuedRangeUpdate = YES;
+    dispatch_async(dispatch_get_main_queue(), ^{
+      [self _updateVisibleNodeIndexPaths];
+    });
   }
-  
-  // coalesce these events -- handling them multiple times per runloop is noisy and expensive
-  _queuedRangeUpdate = YES;
-  
-  dispatch_async(dispatch_get_main_queue(), ^{
-    [self performRangeUpdate];
-  });
-}
-
-- (void)performRangeUpdate
-{
-  // Call this version if you want the update to occur immediately, such as on app suspend, as another runloop may not occur.
-  ASDisplayNodeAssertMainThread();
-  _queuedRangeUpdate = YES; // For now, set this flag as _update... expects it and clears it.
-  [self _updateVisibleNodeIndexPaths];
 }
 
 - (void)setLayoutController:(id<ASLayoutController>)layoutController
 {
   _layoutController = layoutController;
   _layoutControllerImplementsSetVisibleIndexPaths = [_layoutController respondsToSelector:@selector(setVisibleNodeIndexPaths:)];
   if (_layoutController && _queuedRangeUpdate) {
-    [self performRangeUpdate];
+    [self performRangeUpdateSynchronously:YES];
   }
 }
 
 - (void)setDataSource:(id<ASRangeControllerDataSource>)dataSource
 {
   _dataSource = dataSource;
   if (_dataSource && _queuedRangeUpdate) {
-    [self performRangeUpdate];
+    [self performRangeUpdateSynchronously:YES];
   }
 }
 
 - (void)_updateVisibleNodeIndexPaths
 {
   ASDisplayNodeAssert(_layoutController, @"An ASLayoutController is required by ASRangeController");
-  if (!_queuedRangeUpdate || !_layoutController || !_dataSource) {
+  if (!_layoutController || !_dataSource) {
```

Comment:
I believe this is handled in 14842b5 . The only code path that reaches `_updateVisibleNodeIndexPaths` is the queueing method `performRangeUpdateSynchronously:` and it now correctly checks/sets the queued flag.